File: dataflow-java/Sideinputs.java
// Java - Execute a Dataflow pipeline that can carry out Map and Reduce operations, use side inputs, and stream into BigQuery.

// step1: git clone https://github.com/GoogleCloudPlatform/training-data-analyst

// step2: Dataflow APi is succesfully enabled

// gcloud services disable dataflow.googleapis.com --force
// gcloud services enable dataflow.googleapis.com

// step3: set environment variables

// PROJECT="qwiklabs-gcp-02-31fcb8c43a7a"
// echo $PROJECT

// BUCKET="qwiklabs-gcp-02-31fcb8c43a7a"
// echo $BUCKET

// REGION="us-east1"
// echo $REGION

// step4: How to execute the code?
// cd ~/training-data-analyst/courses/data_analysis/lab2/javahelp
// ./run_oncloud3.sh $PROJECT $BUCKET JavaProjectsThatNeedHelp $REGION

// step5: Test the output
// BUCKET="qwiklabs-gcp-02-31fcb8c43a7a"
// gcloud storage cp gs://$BUCKET/javahelp/output.csv .
// head output.csv

// step6: Code

package com.google.cloud.training.dataanalyst.javahelp;

import java.util.ArrayList;
import java.util.List;

import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.io.TextIO;
import org.apache.beam.sdk.options.Default;
import org.apache.beam.sdk.options.Description;
import org.apache.beam.sdk.options.PipelineOptions;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.ParDo;
import org.apache.beam.sdk.transforms.Sum;
import org.apache.beam.sdk.transforms.Top;
import org.apache.beam.sdk.values.KV;

/**
 * A dataflow pipeline that finds the most commonly imported packages
 * 
 * @author vlakshmanan
 *
 */
public class IsPopular {

	public static interface MyOptions extends PipelineOptions {
		@Description("Output prefix")
		@Default.String("/tmp/output")
		String getOutputPrefix();

		void setOutputPrefix(String s);
		
		@Description("Input directory")
		@Default.String("src/main/java/com/google/cloud/training/dataanalyst/javahelp/")
		String getInput();

		void setInput(String s);
	}
	
	@SuppressWarnings("serial")
	public static void main(String[] args) {
		MyOptions options = PipelineOptionsFactory.fromArgs(args).withValidation().as(MyOptions.class);
		Pipeline p = Pipeline.create(options);

		String input = options.getInput() + "*.java";
		String outputPrefix = options.getOutputPrefix();
		final String keyword = "import";
		
		p //
				.apply("GetJava", TextIO.read().from(input)) //
				.apply("GetImports", ParDo.of(new DoFn<String, String>() {
					@ProcessElement
					public void processElement(ProcessContext c) throws Exception {
						String line = c.element();
						if (line.startsWith(keyword)) {
							c.output(line);
						}
					}
				})) //
				.apply("PackageUse", ParDo.of(new DoFn<String, KV<String,Integer>>() {
					@ProcessElement
					public void processElement(ProcessContext c) throws Exception {
						List<String> packages = getPackages(c.element(), keyword);
						for (String p : packages) {
							c.output(KV.of(p, 1));
						}
					}
				})) //
				.apply(Sum.integersPerKey())
				.apply("Top_5", Top.of(5, new KV.OrderByValue<>())) //
				.apply("ToString", ParDo.of(new DoFn<List<KV<String, Integer>>, String>() {

					@ProcessElement
					public void processElement(ProcessContext c) throws Exception {
						StringBuffer sb = new StringBuffer();
						for (KV<String, Integer> kv : c.element()) {
							sb.append(kv.getKey() + "," + kv.getValue() + '\n');
						}
						c.output(sb.toString());
					}

				})) //
				.apply(TextIO.write().to(outputPrefix).withSuffix(".csv").withoutSharding());

		p.run();
	}
	
	private static List<String> getPackages(String line, String keyword) {
		int start = line.indexOf(keyword) + keyword.length();
		int end = line.indexOf(";", start);
		if (start < end) {
			String packageName = line.substring(start, end).trim();
			return splitPackageName(packageName);
		}
		return new ArrayList<String>();
	}
	
	private static List<String> splitPackageName(String packageName) {
		// e.g. given com.example.appname.library.widgetname
		// returns com
		// com.example
		// com.example.appname
		// etc.
		List<String> result = new ArrayList<>();
		int end = packageName.indexOf('.');
		while (end > 0) {
			result.add(packageName.substring(0, end));
			end = packageName.indexOf('.', end + 1);
		}
		result.add(packageName);
		return result;
	}
}

File: dataflow-python/sideInputs.py
# * Read data from BigQuery into Dataflow
# * Use the output of a pipeline as a side-input to another pipeline

# step1: grant dataflow developer role to compute engine service account

# step2: enable dataflow api
# gcloud services disable dataflow.googleapis.com
# gcloud services enable dataflow.googleapis.com

# step3: git clone https://github.com/GoogleCloudPlatform/training-data-analyst

# step4: Create a bucket with same name as project_ID

# step5: in terminal

# BUCKET="qwiklabs-gcp-04-e54235ab2ec7"
# echo $BUCKET

# PROJECT="qwiklabs-gcp-04-e54235ab2ec7"
# echo $PROJECT

# REGION="us-central1"
# echo $REGION

# step6: Query in bigquery

# SELECT
#   content
# FROM
#   `cloud-training-demos.github_repos.contents_java`
# LIMIT
#   10
  
# The BigQuery table cloud-training-demos.github_repos.contents_java contains the content
#  (and some metadata) of all the Java files present in GitHub in 2016.
 
# step6: 
# cd ~/training-data-analyst/courses/data_analysis/lab2/python

# nano JavaProjectsThatNeedHelp.py

# step7: How to run the program on cloud

# python3 JavaProjectsThatNeedHelp.py --bucket $BUCKET --project $PROJECT --region $REGION --DataFlowRunner

# step8: How to run the program locally

# python3 JavaProjectsThatNeedHelp.py --bucket $BUCKET --project $PROJECT --region $REGION --DirectRunner

# step9: Code


"""
Copyright Google Inc. 2018
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import argparse
import logging
import datetime, os
import apache_beam as beam
import math

'''

This is a dataflow pipeline that demonstrates Python use of side inputs. The pipeline finds Java packages
on Github that are (a) popular and (b) need help. Popularity is use of the package in a lot of other
projects, and is determined by counting the number of times the package appears in import statements.
Needing help is determined by counting the number of times the package contains the words FIXME or TODO
in its source.

@author tomstern
based on original work by vlakshmanan

python JavaProjectsThatNeedHelp.py --project <PROJECT> --bucket <BUCKET> --DirectRunner or --DataFlowRunner

'''

# Global values
TOPN=1000


# ### Functions used for both main and side inputs

def splitPackageName(packageName):
   """e.g. given com.example.appname.library.widgetname
           returns com
                   com.example
                   com.example.appname
      etc.
   """
   result = []
   end = packageName.find('.')
   while end > 0:
      result.append(packageName[0:end])
      end = packageName.find('.', end+1)
   result.append(packageName)
   return result

def getPackages(line, keyword):
   start = line.find(keyword) + len(keyword)
   end = line.find(';', start)
   if start < end:
      packageName = line[start:end].strip()
      return splitPackageName(packageName)
   return []

def packageUse(record, keyword):
   if record is not None:
     lines=record.split('\n')
     for line in lines:
       if line.startswith(keyword):
         packages = getPackages(line, keyword)
         for p in packages:
		    yield (p, 1)

def is_popular(pcoll):
 return (pcoll
    | 'PackageUse' >> beam.FlatMap(lambda rowdict: packageUse(rowdict['content'], 'import'))
    | 'TotalUse' >> beam.CombinePerKey(sum)
    | 'Top_NNN' >> beam.transforms.combiners.Top.Of(TOPN, key=lambda kv: kv[1]) )



def packageHelp(record, keyword):
   count=0
   package_name=''
   if record is not None:
     lines=record.split('\n')
     for line in lines:
       if line.startswith(keyword):
         package_name=line
       if 'FIXME' in line or 'TODO' in line:
         count+=1
     packages = (getPackages(package_name, keyword) )
     for p in packages:
         yield (p,count)

def needs_help(pcoll):
 return (pcoll
    | 'PackageHelp' >> beam.FlatMap(lambda rowdict: packageHelp(rowdict['content'], 'package'))
    | 'TotalHelp' >> beam.CombinePerKey(sum)
    | 'DropZero' >> beam.Filter(lambda packages: packages[1]>0 ) )


# Calculate the final composite score
#
#    For each package that is popular
#    If the package is in the needs help dictionary, retrieve the popularity count
#    Multiply to get compositescore
#      - Using log() because these measures are subject to tournament effects
#

def compositeScore(popular, help):
    for element in popular:
      if help.get(element[0]):
         composite = math.log(help.get(element[0])) * math.log(element[1])
         if composite > 0:
           yield (element[0], composite)


# ### main

# Define pipeline runner (lazy execution)
def run():

  # Command line arguments
  parser = argparse.ArgumentParser(description='Demonstrate side inputs')
  parser.add_argument('--bucket', required=True, help='Specify Cloud Storage bucket for output')
  parser.add_argument('--project',required=True, help='Specify Google Cloud project')
  parser.add_argument('--region',required=True, help='Specify Google Cloud region')
  group = parser.add_mutually_exclusive_group(required=True)
  group.add_argument('--DirectRunner',action='store_true')
  group.add_argument('--DataFlowRunner',action='store_true')

  opts = parser.parse_args()

  if opts.DirectRunner:
    runner='DirectRunner'
  if opts.DataFlowRunner:
    runner='DataFlowRunner'

  bucket = opts.bucket
  project = opts.project
  region = opts.region

  #    Limit records if running local, or full data if running on the cloud
  limit_records=''
  if runner == 'DirectRunner':
     limit_records='LIMIT 3000'
  get_java_query='SELECT content FROM [cloud-training-demos:github_repos.contents_java] {0}'.format(limit_record>

  argv = [
    '--project={0}'.format(project),
    '--job_name=javahelpjob',
    '--save_main_session',
    '--worker_machine_type=e2-standard-2',
    '--staging_location=gs://{0}/staging/'.format(bucket),
    '--temp_location=gs://{0}/staging/'.format(bucket),
    '--runner={0}'.format(runner),
    '--region={0}'.format(region),
    '--max_num_workers=5'
    ]

  p = beam.Pipeline(argv=argv)


  # Read the table rows into a PCollection (a Python Dictionary)
  bigqcollection = p | 'ReadFromBQ' >> beam.io.Read(beam.io.BigQuerySource(project=project,query=get_java_query))

  popular_packages = is_popular(bigqcollection) # main input

  help_packages = needs_help(bigqcollection) # side input

  # Use side inputs to view the help_packages as a dictionary
  results = popular_packages | 'Scores' >> beam.FlatMap(lambda element, the_dict: compositeScore(element,the_dic>

  # Write out the composite scores and packages to an unsharded csv file
  output_results = 'gs://{0}/javahelp/Results'.format(bucket)
   results | 'WriteToStorage' >> beam.io.WriteToText(output_results,file_name_suffix='.csv',shard_name_template='>

  # Run the pipeline (all operations are deferred until run() is called).


  if runner == 'DataFlowRunner':
     p.run()
  else:
     p.run().wait_until_finish()
  logging.getLogger().setLevel(logging.INFO)


if __name__ == '__main__':
  run()
